cclunit-framework-source/src/main/resources/cclutmock.inc

```diff
@@ -0,0 +1,192 @@
+%#ifndef CCLUTMOCK_INC
+%#def CCLUTMOCK_INC
+;**********************************************************************************************************************************
+;
+; Provides support functions that can be leveraged by consumers to construct mock objects and use those mocks in the script under
+; test.
+;
+;**********************************************************************************************************************************
+
+%i cclsource:cclutmock_table.inc
+
+declare cclutAddMockImplementation(originalName = vc(val), replaceName = vc(val)) = null with protect
+declare cclutRemoveMockImplementation(originalName = vc(val)) = null with protect
+declare cclutRemoveAllMockImplementations(null) = null with protect
+declare cclutExecuteProgramWithMocks(programName = vc(val), params = vc(val),
+    namespace = vc(val, "PUBLIC")) = null with protect
+declare cclutRemoveAllMocks(null) = null with protect
+
+record cclut_mockImplementations (
+    1 implementations[*]
+		2 originalName = vc
+		2 replaceName = vc
+) with protect
+
+/**
+Adds a mock implementation to be utilized by cclutExecuteProgramWithMocks.  This can include record structures,
+subroutines, or other programs.  originalName and replaceName are required.  If originalName is already being mocked,
+the mock will be overwritten with the new replaceName.
+
+@param originalName
+    The object to be mocked.
+@param replaceName
+    The mocked object.
+
+Example:
+call cclutAddMockImplementation("uar_get_code_by", "mock_uar_get_code_by")
+**/
+subroutine cclutAddMockImplementation(originalName, replaceName)
+    ;Declare local variables
+    declare CCLUT_ORIGINAL_NAME = vc with protect, constant(trim(cnvtupper(originalName), 3))
```

Comment:
Is it necessary?  Not particularly.  But in general with CCL, it's usually a common convention to do some mild cleaning of data to process instead of being extremely rigid.  I've worked with plenty of scripts that do this, and the CCL Unit Testing framework already has patterns like this.  For example, the cclutTimerName is trimmed before it's used.

https://github.com/cerner/cclunit-framework/blob/master/cclunit-framework-source/src/main/resources/cclut_timers.inc#L44

Reply:
I advocate understanding what is necessary (and eliminating what is not) opposed to blindly following a practice that lacks motivation just because it is done that way in lots of other places.   If we really believe we should support coding constructs like **call ccultAddMockImplementation("       some_thing", "                     mock_some_thing")**, then the trim should stay.  Personally, I don't think we should and I am willing to address all the support questions that arise because of that choice.  I do not disagree that there are still areas of the existing framework code that could be improved and/or cleaned up.

Reply:
I don't consider it a blind practice (at least, that's not why I do it).  There's a very real motivation to it - being flexible enough to handle minor edge cases.

The example you have would be something I believe we should support.  Since you feel that the timer code I linked to is also not desirable, I assume showing other examples of Cerner's code base doing this would also be insufficient, so I'll stick to more abstract examples.

One example might be of someone who has a file that they're parsing through and these implementations (and mocks) are in a loop.  Depending on how they've structured their code, they might have some leftover whitespace in front/after the string being parsed.

Setting aside the inconsistencies from the fact that
call ccultAddMockImplementation(" some_thing", " mock_some_thing")
wouldn't work, while
call ccultAddMockImplementation("some_thing ", "mock_some_thing ")
would (since that's part of the language construct, the API consumer should be familiar with it), I favor flexible APIs that can make reasonable assumptions (nothing with a space would be mockable since CCL names can't have them, so trimming is a reasonable assumption).

To say that we _don't_ want that would imply to me that we want to ensure our consumers are writing very exact code.  In my example above with the file, could we say "The consumer should trim their names before sending them in"?  Of course.  But then, there are other aspects we should question as well.

For example, I am allowing the names to be case-insensitive by normalizing them to uppercase.  The below would be permissible currently:
call cclutDefineMockTable("person", "person_id", "f8")
call cclutCreateMockTable("PERSON")

This is also not strictly necessary (we could validate the case and throw errors if the casing does not match what was used in the first call), but that seems overbearing when we can accept the reasonable likelihood that they are referring to the same table (or excluding whitespace).

If it was included as an implementation note (that whitespace will be trimmed), would that make it more palatable?  Or are you against it regardless?

Reply:
The documentation says "The object to be mocked." not "Something that when trimmed becomes the object to be mocked."

A blanket statement about trimming could be used to make the first statement imply the second but....
I would not classify "we promise to always trim everything" the same as "we promise to always ignore case" in a case-insensitive language with nonsensical string handling behavior.  

I also think there are places (adding mock data) where we should definitely not do any trimming. 

I didn't intend for "blindly following" to be a harsh comment.  It's simply that the initial justification for retaining the trimming was "...it's usually a common convention..." and "...plenty of scripts that do this..." which sounds to me like "because everyone else does".  Every case is different. Is there a solid reason why we should provide trimming here?

The provided example is not. It just sounds like sloppy coding to me.  The parsing code should be written to extract exactly what is needed and it should be checked/tested to ensure that it does. The consumer can accomplish this feat without doing any trimming by pulling the correct substring, but the consumer is also welcome to extract then trim away some extra white space if that is the only way they can find.

Reply:
> I would not classify "we promise to always trim everything" the same as "we promise to always ignore case" in a case-insensitive language with nonsensical string handling behavior.
I also think there are places (adding mock data) where we should definitely not do any trimming.

Internally with the mocking API, we're (or at least I intended to be) consistent about this.

I fully agree that cclutAddMockData should not do trimming (nor does it currently) as it should be an exact representation of the consumer's entry (no reasonable assumptions can be made here since it's table data).  Additionally, it does no case manipulation (nor should it) under the same principle.  However, while CCL is case-insensitive for syntax, it is not for strings (e.g. "person" != "PERSON")

My purpose in mentioning that is _if_ we remove the trimming (which I still disagree with), it feels inconsistent to leave case manipulation as well.  In other words, for these lines:
call cclutDefineMockTable("person", "person_id", "f8")
call cclutCreateMockTable("PERSON")

I would expect the second line to throw an error that "PERSON" has not been defined under a similar idea of sloppy coding (i.e. they should have used lowercase again).  However, to me, I view it as flexibility under reasonable assumptions rather than trying to sweep up poor coding practices.  I think this could still be documented without a "full" blanket statement by saying that "names" (i.e. subroutines, tables, record_structures, etc.) are trimmed.  This would then exempt things like "params" from cclutExecuteProgramWithMocks or "rowData" from cclutAddMockData.

Said another way, I'm not trying to "push" what is/isn't sloppy on to the consumer.  I'm treating our API as the system with the person writing the test as a consumer, so if it's easy enough to support and can save potential headache, that seems like a good thing to me.  Similarly, if I'm on GitHub using their Search bar to search for " cclunit-framework", it still pulls up this repository.  Now, I know that is a _huge_ stretch because I expect programmers to be a cut above the average user in terms of precision, and the GitHub search is a front-end UI control, but my point is more an abstract one regarding the system/consumer relationship.

> I didn't intend for "blindly following" to be a harsh comment. It's simply that the initial justification for retaining the trimming was "...it's usually a common convention..." and "...plenty of scripts that do this..." which sounds to me like "because everyone else does". Every case is different. Is there a solid reason why we should provide trimming here?

My apologies as I did not mean for it to be a "because everyone else does" statement.  My intent was to demonstrate that it's a common cleaning mechanism for inbound data where the whitespaces are not relevant (and I agree with its usage in the cclutTimers even though it sounds like you'll potentially take that out in the future).

Reply:
> it feels inconsistent to leave case manipulation

Why do you find it strange to ignore case but not white space?  CCL is case-insensitive, not white-space-insensitive.   In fact I find it white-space-make-you-want-to-puke-sensitive.
```
    declare str1 = vc with protect, nonconstant("") ;;; length? ONE - CHAR(0)
    decare str2 = vc with protect, noconstant(" ")  ;;; length? ZERO
    set str1 = concat(trim(str1), "more") ;;; without the trim textlen(str1) is 5 - initial character CHAR(0)
    set str2 = concat(str2, "more") ;;; str2 ends up being "more"
    set str2 = "i have trailing whitespace    " ;; str2 ends up being "i have trailing whitespace"
    set str2 = "     i have leading whitespace" ;; str2 ends up being "    i have trailing whitespace"
```

We are trying to help engineers achieve a higher standard in production code.  Why not expect the same from their unit tests.  I think trimming will ultimately lead to additional confusion about how CCL handles white space more than it will lead to anything else.  Realize we are not creating a production HCIT application that needs to be resilient to bad data.

Some long time ago Cerner experienced "the long data sequence problem". Someone wrote a piece of code something like **set order_id = cnvtlong(seq(order_seq, nextval))**.  Someone modeled that insert statement and someone modeled that one and.... till pretty much every database insert in Cerner's code base looked just like that.  Eventually there was an SR because the next id that came out of the bucket in a production environment overflowed the size of a long and the inserted id ended up being negative number (fortunately not a duplicate).  All new development ceased at Cerner for about two months while teams identified and repaired instances of this defect. 

Let's use a better yardstick than "it is common".



I can't find the unit tests which fail if this trimming is removed. Can you point them out?
 
We needn't belabor cclutAddMockData.  I only mentioned it because it precludes adding a blanket statement that all inputs will be automatically trimmed.

Reply:
> Why do you find it strange to ignore case but not white space? CCL is case-insensitive, not white-space-insensitive. In fact I find it white-space-make-you-want-to-puke-sensitive.

As I mentioned, CCL is only case-insensitive for syntax when writing source code, not for strings.  That is why I find it strange.  It's still a level of manipulation different than what the user supplied to the function.  I understand CCL's irregularities when it comes to strings and whitespaces, but I don't feel that's terribly relevant to my point (as I'd still want to clean trailing whitespace even if CCL didn't already do it).

> We are trying to help engineers achieve a higher standard in production code. Why not expect the same from their unit tests. I think trimming will ultimately lead to additional confusion about how CCL handles white space more than it will lead to anything else. Realize we are not creating a production HCIT application that needs to be resilient to bad data.

This also seems inconsistent to me.  You're indicating that we want to help engineers achieve a higher standard in production code, and to expect that from their unit tests.  If so, then why shouldn't _we_ expect that from the very framework they're using to test it?  If we want the framework to be an example of desirable code and coding standards, then I do think it's important that we code to "production" level.  For that reason, I think that doing some basic error checking (such as checking for empty values, which we do) and massaging of data where appropriate and sensible (which I believe is the case for trimming table/object names) meets that level.

Alternatively, to use the reverse argument.  If we're saying that it's not good quality to have leading spaces (which I do agree that it's not), then I would say that it's not good quality to use differing cases in the same test, such as my earlier example:
call cclutDefineMockTable("person", "person_id", "f8")
call cclutCreateMockTable("PERSON")

There's certainly no reason that the engineer should need to make the second "PERSON" uppercase, when it should be consistent with the casing used in their first statement.

> Let's use a better yardstick than "it is common".

Fair enough.  That was only one of my points; not the main one.  My main reason in choosing this was a way of cleaning the data with relative ease where a reasonable assumption could be made.

> I can't find the unit tests which fail if this trimming is removed. Can you point them out?

If I write one, will you change your mind? 😄 Sorry, don't mean to be facetious, but yes, I should have written some tests that incorporate it.  However, similarly, I don't have tests that would fail due to the case manipulation either (that's why I don't feel it's a good question; my failure to account for all scenarios shouldn't be a reason to include/exclude something).

> We needn't belabor cclutAddMockData. I only mentioned it because it precludes adding a blanket statement that all inputs will be automatically trimmed.

My original mention of adding the trimming note to the Implementation Notes can come with whatever qualifications/criteria we feel are appropriate.  I didn't originally intend for it to be a "blanket" statement (my apologies if it came off that way).  I was just evaluating if that would be an acceptable compromise to leaving in the trimming.

Reply:
> If I write one, will you change your mind.

Not really.  The trim does nothing but allow a unit test to pass in parameters having leading white space.  I see no value in allowing that nor in wasting time documenting that it is or is not supported.

I mentioned the lack of unit tests because it suggests the provided 'feature' was less than intentional.

There are many, many production-quality functions which do not allow leading/trailing white space to be included willy-nilly in their input string parameters.  I would actually argue few if any functions from programming languages or programming frameworks do this outside of ones that convert strings to dates or numbers.  Here are a handful of the many examples that do not.
```
uar_get_code_by("    MEANING    ",  8, "    ACTIVE    ")
findstring("what I am looking for is in here somewhere", "    what I am looking for     ")
tdbexecute(3202004, 3202004, 4410708, "   REC",   req,  "   REC   ",  rep) 

document.getElementById('   someElementId  ');
```
Any maven command or maven plugin command (I cannot attest definitively to the latter but I leave you to find a counter example). Any Windows API call.

Reply:
I've said before that I still disagree, but I don't think there's any point furthering the discussion as it doesn't look like either of us will be convinced.  I concede the point and will make the change everywhere.